### PR TITLE
save multi ABCIResponses with related height by batch

### PR DESCRIPTION
**Pull request by mmsqe**
This pr allow update db without unnecessary Marshal with [exported](https://github.com/crypto-org-chain/cronos/blob/21dc070b63d2f54a774cb74aedd8c6351f01f7e7/cmd/cronosd/cmd/fix-unlucky-tx.go#L199) abci response raw data, and [faster](https://github.com/crypto-org-chain/cronos/pull/588#discussion_r923987828) with WriteSync in batch compare with SetSync one by one.

### CODE REVIEW

**Comment by cmwaters** [comment]
Hi @mmsqe, can you provide a description of the PR and what problem you are trying to resolve.

File: state/store.go [hunk 1]
```diff
@@ -64,6 +64,8 @@ type Store interface {
 	Save(State) error
 	// SaveABCIResponses saves ABCIResponses for a given height
 	SaveABCIResponses(int64, *tmstate.ABCIResponses) error
+	// SaveBatchABCIResponseBytes saves multi ABCIResponses with related height by batch
```
Comment (JayT106) [inline state/store.go:67]:
```suggestion
	// SaveBatchABCIResponseBytes saves a batch of ABCIResponses with given heights
```

File: state/store.go [hunk 2]
```diff
@@ -414,6 +416,27 @@ func (store dbStore) SaveABCIResponses(height int64, abciResponses *tmstate.ABCI
 	return nil
 }
 
+// SaveBatchABCIResponseBytes persists heights with ABCIResponses to the database by batch.
+func (store dbStore) SaveBatchABCIResponseBytes(heights []int64, abciResponseBytes [][]byte) error {
+	heightLen := len(heights)
+	bytesLen := len(abciResponseBytes)
+	if heightLen != bytesLen {
+		return fmt.Errorf("heights doesn't match abciResponseBytes: %d vs %d",
```
Comment (JayT106) [inline state/store.go:424]:
```suggestion
		return fmt.Errorf("SaveBatchABCIResponseBytes: the size of height %d doesn't match with the size of abciResponseBytes %v",
```

File: state/store.go [hunk 3]
```diff
@@ -414,6 +416,27 @@ func (store dbStore) SaveABCIResponses(height int64, abciResponses *tmstate.ABCI
...
+			bytesLen,
+		)
+	}
+	batch := store.db.NewBatch()
+	defer batch.Close()
+	for i, height := range heights {
+		err := batch.Set(calcABCIResponsesKey(height), abciResponseBytes[i])
+		if err != nil {
+			return err
+		}
```
Comment (JayT106) [inline state/store.go:435]:
```suggestion
```suggestion
		if err := batch.Set(calcABCIResponsesKey(height), abciResponseBytes[i]); err != nil {
			return err
		}
```
```

**Comment by cmwaters** [comment]
Thanks for the description. From where are you planning on saving multiple ABCIResponses at once?

**Comment by JayT106** [comment]
> Thanks for the description. From where are you planning on saving multiple ABCIResponses at once?

It's PR in Cronos
https://github.com/crypto-org-chain/cronos/pull/588

**Comment by williambanfield** [comment]
@mmsqe could you please provide some metrics and benchmarks demonstrating the improvement?

**Comment by mmsqe** [comment]
> benchmarks

add some simple test, but prod has bigger diff with more complex data
```
go test -bench=BenchmarkSaveBatchABCIResponseBytes -count 5
goos: darwin
goarch: amd64
pkg: github.com/tendermint/tendermint/state
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkSaveBatchABCIResponseBytes/height=100-16         	   36763	     31875 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=100-16         	   37533	     31691 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=100-16         	   38786	     31455 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=100-16         	   36078	     31414 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=100-16         	   35607	     31888 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=1000-16        	    3277	    349348 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=1000-16        	    3400	    346981 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=1000-16        	    3105	    382339 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=1000-16        	    3031	    342342 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=1000-16        	    3451	    351751 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=10000-16       	     280	   4223164 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=10000-16       	     285	   4481244 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=10000-16       	     265	   4500882 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=10000-16       	     278	   4406737 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=10000-16       	     264	   4459862 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=100000-16      	      21	  59752316 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=100000-16      	      21	  56612356 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=100000-16      	      21	  57403535 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=100000-16      	      20	  56318762 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=100000-16      	      22	  57991664 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=1000000-16     	       2	 652690224 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=1000000-16     	       2	 629749707 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=1000000-16     	       2	 605627208 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=1000000-16     	       2	 577671462 ns/op
BenchmarkSaveBatchABCIResponseBytes/height=1000000-16     	       2	 579237246 ns/op
PASS
ok  	github.com/tendermint/tendermint/state	75.767s
```
vs
```
go test -bench=BenchmarkSaveABCIResponses -count 5
goos: darwin
goarch: amd64
pkg: github.com/tendermint/tendermint/state
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkSaveABCIResponses/height=100-16         	    2487	    465309 ns/op
BenchmarkSaveABCIResponses/height=100-16         	    2668	    445390 ns/op
BenchmarkSaveABCIResponses/height=100-16         	    2430	    442723 ns/op
BenchmarkSaveABCIResponses/height=100-16         	    2697	    452913 ns/op
BenchmarkSaveABCIResponses/height=100-16         	    2572	    467269 ns/op
BenchmarkSaveABCIResponses/height=1000-16        	     241	   5033293 ns/op
BenchmarkSaveABCIResponses/height=1000-16        	     243	   5212257 ns/op
BenchmarkSaveABCIResponses/height=1000-16        	     212	   4942610 ns/op
BenchmarkSaveABCIResponses/height=1000-16        	     232	   5027986 ns/op
BenchmarkSaveABCIResponses/height=1000-16        	     240	   5082715 ns/op
BenchmarkSaveABCIResponses/height=10000-16       	      19	  54447997 ns/op
BenchmarkSaveABCIResponses/height=10000-16       	      22	  55018674 ns/op
BenchmarkSaveABCIResponses/height=10000-16       	      21	  54534192 ns/op
BenchmarkSaveABCIResponses/height=10000-16       	      21	  56078144 ns/op
BenchmarkSaveABCIResponses/height=10000-16       	      21	  52921121 ns/op
BenchmarkSaveABCIResponses/height=100000-16      	       2	 762133067 ns/op
BenchmarkSaveABCIResponses/height=100000-16      	       2	 595150425 ns/op
BenchmarkSaveABCIResponses/height=100000-16      	       2	 546551444 ns/op
BenchmarkSaveABCIResponses/height=100000-16      	       2	 555613772 ns/op
BenchmarkSaveABCIResponses/height=100000-16      	       2	 549713295 ns/op
BenchmarkSaveABCIResponses/height=1000000-16     	       1	8529069490 ns/op
BenchmarkSaveABCIResponses/height=1000000-16     	       1	6999325788 ns/op
BenchmarkSaveABCIResponses/height=1000000-16     	       1	5573764997 ns/op
BenchmarkSaveABCIResponses/height=1000000-16     	       1	5732692024 ns/op
BenchmarkSaveABCIResponses/height=1000000-16     	       1	5632021911 ns/op
PASS
ok  	github.com/tendermint/tendermint/state	108.864s
```

**Comment by cmwaters** [comment]
Thanks for all the context - it's been helpful. I guess the assessment I'm trying to make here is, is this feature something that belongs on mainline tendermint. Your team can of course implement it in your own repo. It wouldn't be too hard to use tm-db to create the function of batch saving abci response bytes. I just want to know if you think a lot of users would benefit from this.

**Comment by mmsqe** [comment]
> Thanks for all the context - it's been helpful. 
yes, I think we're not alone that need this change, but we could close this pr since no other use case for now.

**Pull request closed without merging**